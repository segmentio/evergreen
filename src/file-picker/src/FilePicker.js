import React, { PureComponent } from 'react'
import PropTypes from 'prop-types'
import Box from 'ui-box'
import { Button } from '../../buttons'
import { TextInput } from '../../text-input'

export const CLASS_PREFIX = 'evergreen-file-picker'

export default class FilePicker extends PureComponent {
  static propTypes = {
    /**
     * Name attribute of the input.
     */
    name: PropTypes.string,

    /**
     * The accept attribute of the input.
     */
    accept: PropTypes.oneOfType([
      PropTypes.string,
      PropTypes.arrayOf(PropTypes.string)
    ]),

    /**
     * When true, the file picker is required.
     */
    required: PropTypes.bool,

    /**
     * When true, accept multiple files.
     */
    multiple: PropTypes.bool,

    /**
     * When true, the filepicker is disabled.
     */
    disabled: PropTypes.bool,

    /**
     * The capture attribute of the input.
     */
    capture: PropTypes.bool,

    /**
     * The height of the file picker.
     */
    height: PropTypes.number,

    /**
     * Function called when onChange is fired
     */
    onChange: PropTypes.func
  }

  constructor() {
    super()

    this.state = {
      files: []
    }
  }

  render() {
    const {
      name,
      accept,
      required,
      multiple,
      disabled,
      capture,
      height,
      onChange, // Remove onChange from props
      ...props
    } = this.props
    const { files } = this.state

    let inputValue
    if (files.length === 0) {
      inputValue = ''
    } else if (files.length === 1) {
      inputValue = files[0].name
    } else {
      inputValue = `${files.length} files`
    }

    let buttonText
    if (files.length === 0) {
      buttonText = 'Select file'
    } else if (files.length === 1) {
      buttonText = 'Replace file'
    } else {
      buttonText = 'Replace files'
    }

    return (
      <Box display="flex" className={`${CLASS_PREFIX}-root`} {...props}>
        <Box
          innerRef={this.fileInputRef}
          className={`${CLASS_PREFIX}-file-input`}
          is="input"
          type="file"
          name={name}
          accept={accept}
          required={required}
          multiple={multiple}
          disabled={disabled}
          capture={capture}
          onChange={this.handleFileChange}
          display="none"
        />

        <TextInput
          className={`${CLASS_PREFIX}-text-input`}
          readOnly
          value={inputValue}
          placeholder="Select a file to uploadâ€¦"
          // There's a weird specifity issue when there's two differently sized inputs on the page
          borderTopRightRadius="0 !important"
          borderBottomRightRadius="0 !important"
          height={height}
          flex={1}
          textOverflow="ellipsis"
        />

        <Button
          className={`${CLASS_PREFIX}-button`}
          onClick={this.handleButtonClick}
          disabled={disabled}
          borderTopLeftRadius={0}
          borderBottomLeftRadius={0}
          height={height}
          flexShrink={0}
          type="button"
        >
          {buttonText}
        </Button>
      </Box>
    )
  }

  fileInputRef = node => {
    this.fileInput = node
  }

  handleFileChange = e => {
    const { onChange } = this.props
    // Firefox returns the same array instance each time for some reason
    const files = [...e.target.files]

    this.setState({ files })

    if (onChange) {
      onChange(files)
    }
  }

  handleButtonClick = () => {
    this.fileInput.click()
  }
}
