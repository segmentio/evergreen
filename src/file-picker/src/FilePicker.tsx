import React, { memo, forwardRef, useState, useRef, useCallback } from 'react'
import cx from 'classnames'
import Box, { PolymorphicBoxProps } from 'ui-box'
import { Button } from '../../buttons'
import safeInvoke from '../../lib/safe-invoke'
import { TextInput } from '../../text-input'

export interface FilePickerOwnProps {
    /** the name attribute of the input */
    name?: string;
    /** the accept attribute of the input */
    accept?: string | string[];
    /** whether or not the field is required */
    required?: boolean;
    /** whether or not the file input accepts multiple files */
    multiple?: boolean;
    /** whether or not the filepicker is disabled */
    disabled?: boolean;
    /** the capture attribute of the input */
    capture?: boolean;
    /** the height of the filepicker */
    height?: number;
    /** function called when onChange is fired */
    onChange?: (files: FileList) => void;
    /** function called when onBlur is fired */
    onBlur?: (event: React.FocusEvent) => void;
    /** placeholder of the text input */
    placeholder?: string;
}

export type FilePickerProps = PolymorphicBoxProps<'div', FilePickerOwnProps>;

export const CLASS_PREFIX = 'evergreen-file-picker'

const FilePicker: React.FC<FilePickerProps> = memo(
  forwardRef(function FilePicker(props, ref) {
    const {
      accept,
      capture,
      className,
      disabled,
      height = 32,
      multiple,
      name,
      onBlur,
      onChange,
      placeholder = 'Select a file to uploadâ€¦',
      required,
      ...rest
    } = props

    const [files, setFiles] = useState([])
    const fileInputRef = useRef()

    const handleFileChange = useCallback(
      e => {
        // Firefox returns the same array instance each time for some reason
        const filesCopy = [...e.target.files]

        // @ts-expect-error ts-migrate(2345) FIXME: Argument of type 'any[]' is not assignable to para... Remove this comment to see the full error message
        setFiles(filesCopy)

        safeInvoke(onChange, filesCopy)
      },
      [onChange]
    )

    const handleButtonClick = useCallback(() => {
      if (fileInputRef.current) {
        // @ts-expect-error ts-migrate(2532) FIXME: Object is possibly 'undefined'.
        fileInputRef.current.click()
      }
    }, [])

    const handleBlur = useCallback(
      e => {
        // Setting e.target.files to an array fails. It must be a FileList
        if (e && e.target) {
          // @ts-expect-error ts-migrate(2532) FIXME: Object is possibly 'undefined'.
          e.target.files = fileInputRef.current && fileInputRef.current.files
        }

        safeInvoke(onBlur, e)
      },
      [onBlur]
    )

    let inputValue
    if (files.length === 0) {
      inputValue = ''
    } else if (files.length === 1) {
      // @ts-expect-error ts-migrate(2339) FIXME: Property 'name' does not exist on type 'never'.
      inputValue = files[0].name
    } else {
      inputValue = `${files.length} files`
    }

    let buttonText
    if (files.length === 0) {
      buttonText = 'Select file'
    } else if (files.length === 1) {
      buttonText = 'Replace file'
    } else {
      buttonText = 'Replace files'
    }

    const rootClassNames = cx(`${CLASS_PREFIX}-root`, className)

    return (
      <Box display="flex" className={rootClassNames} ref={ref} {...rest}>
        <Box
          // @ts-expect-error ts-migrate(2322) FIXME: Type 'MutableRefObject<undefined>' is not assignab... Remove this comment to see the full error message
          ref={fileInputRef}
          className={`${CLASS_PREFIX}-file-input`}
          is="input"
          type="file"
          name={name}
          // @ts-expect-error ts-migrate(2322) FIXME: Type 'string | string[] | undefined' is not assign... Remove this comment to see the full error message
          accept={accept}
          required={required}
          multiple={multiple}
          disabled={disabled}
          capture={capture}
          onChange={handleFileChange}
          display="none"
        />

        <TextInput
          className={`${CLASS_PREFIX}-text-input`}
          readOnly
          value={inputValue}
          placeholder={placeholder}
          // There's a weird specifity issue when there's two differently sized inputs on the page
          borderTopRightRadius="0 !important"
          borderBottomRightRadius="0 !important"
          height={height}
          flex={1}
          textOverflow="ellipsis"
          onBlur={handleBlur}
        />

        // @ts-expect-error ts-migrate(2745) FIXME: This JSX tag's 'children' prop expects type 'never... Remove this comment to see the full error message
        // @ts-expect-error ts-migrate(2745) FIXME: This JSX tag's 'children' prop expects type 'never... Remove this comment to see the full error message
        // @ts-expect-error ts-migrate(2745) FIXME: This JSX tag's 'children' prop expects type 'never... Remove this comment to see the full error message
        <Button
          className={`${CLASS_PREFIX}-button`}
          onClick={handleButtonClick}
          disabled={disabled}
          borderTopLeftRadius={0}
          borderBottomLeftRadius={0}
          height={height}
          flexShrink={0}
          type="button"
          onBlur={handleBlur}
        >
          {buttonText}
        </Button>
      </Box>
    )
  })
)

export default FilePicker
