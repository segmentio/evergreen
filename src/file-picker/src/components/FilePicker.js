import React, { PureComponent } from 'react'
import PropTypes from 'prop-types'
import Box from 'ui-box'
import { Button } from 'evergreen-buttons'
import { TextInput } from 'evergreen-text-input'

export const CLASS_PREFIX = 'evergreen-file-picker'

export default class FilePicker extends PureComponent {
  static propTypes = {
    name: PropTypes.string,
    accept: PropTypes.oneOfType([
      PropTypes.string,
      PropTypes.arrayOf(PropTypes.string)
    ]),
    required: PropTypes.bool,
    multiple: PropTypes.bool,
    disabled: PropTypes.bool,
    capture: PropTypes.bool,
    height: PropTypes.number,
    onChange: PropTypes.func
  }

  constructor() {
    super()

    this.state = {
      files: []
    }
  }

  render() {
    const {
      name,
      accept,
      required,
      multiple,
      disabled,
      capture,
      height,
      onChange, // Remove onChange from props
      ...props
    } = this.props
    const { files } = this.state

    let inputValue
    if (files.length === 0) {
      inputValue = ''
    } else if (files.length === 1) {
      inputValue = files[0].name
    } else {
      inputValue = `${files.length} files`
    }

    let buttonText
    if (files.length === 0) {
      buttonText = 'Select file'
    } else if (files.length === 1) {
      buttonText = 'Replace file'
    } else {
      buttonText = 'Replace files'
    }

    return (
      <Box display="flex" className={`${CLASS_PREFIX}-root`} {...props}>
        <Box
          innerRef={this.fileInputRef}
          className={`${CLASS_PREFIX}-file-input`}
          is="input"
          type="file"
          name={name}
          accept={accept}
          required={required}
          multiple={multiple}
          disabled={disabled}
          capture={capture}
          onChange={this.handleFileChange}
          display="none"
        />

        <TextInput
          className={`${CLASS_PREFIX}-text-input`}
          readOnly
          value={inputValue}
          placeholder="Select a file to uploadâ€¦"
          // There's a weird specifity issue when there's two differently sized inputs on the page
          borderTopRightRadius="0 !important"
          borderBottomRightRadius="0 !important"
          height={height}
          flex={1}
          textOverflow="ellipsis"
        />

        <Button
          className={`${CLASS_PREFIX}-button`}
          onClick={this.handleButtonClick}
          disabled={disabled}
          borderTopLeftRadius={0}
          borderBottomLeftRadius={0}
          height={height}
          flexShrink={0}
        >
          {buttonText}
        </Button>
      </Box>
    )
  }

  fileInputRef = node => {
    this.fileInput = node
  }

  handleFileChange = e => {
    const { onChange } = this.props
    // Firefox returns the same array instance each time for some reason
    const files = [...e.target.files]

    this.setState({ files })

    if (onChange) {
      onChange(files)
    }
  }

  handleButtonClick = () => {
    this.fileInput.click()
  }
}
