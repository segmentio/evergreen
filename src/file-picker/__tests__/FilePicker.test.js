import React from 'react'
import { shallow } from 'enzyme'
import render from 'react-test-renderer'

import FilePicker, { CLASS_PREFIX } from '../src/FilePicker'

describe('<FilePicker />', () => {
  it('snapshot', () => {
    const component = <FilePicker />
    const tree = render.create(component).toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('sets name', () => {
    const component = shallow(<FilePicker name="hi" />)
    expect(component.find(`.${CLASS_PREFIX}-file-input`).prop('name')).toEqual('hi')
  })

  it('sets accept', () => {
    const component = shallow(<FilePicker accept="application/json" />)
    expect(component.find(`.${CLASS_PREFIX}-file-input`).prop('accept')).toEqual('application/json')
  })

  it('sets required', () => {
    const component = shallow(<FilePicker required />)
    expect(component.find(`.${CLASS_PREFIX}-file-input`).prop('required')).toBe(true)
  })

  it('sets multiple', () => {
    const component = shallow(<FilePicker multiple />)
    expect(component.find(`.${CLASS_PREFIX}-file-input`).prop('multiple')).toBe(true)
  })

  it('sets disabled', () => {
    const component = shallow(<FilePicker disabled />)
    expect(component.find(`.${CLASS_PREFIX}-file-input`).prop('disabled')).toBe(true)
    expect(component.find(`.${CLASS_PREFIX}-button`).prop('disabled')).toBe(true)
  })

  it('sets capture', () => {
    const component = shallow(<FilePicker capture />)
    expect(component.find(`.${CLASS_PREFIX}-file-input`).prop('capture')).toBe(true)
  })

  it('passes through height', () => {
    const component = shallow(<FilePicker height={20} />)
    expect(component.find(`.${CLASS_PREFIX}-text-input`).prop('height')).toEqual(20)
    expect(component.find(`.${CLASS_PREFIX}-button`).prop('height')).toEqual(20)
  })

  it('passes through props', () => {
    const component = shallow(<FilePicker width={20} />)
    expect(component.find(`.${CLASS_PREFIX}-root`).prop('width')).toEqual(20)
  })

  it('calls onChange', () => {
    const onChange = jest.fn()
    const component = shallow(<FilePicker onChange={onChange} />)
    const e = {
      target: {
        files: [{ name: 'data.json' }]
      }
    }
    component.find(`.${CLASS_PREFIX}-file-input`).simulate('change', e)
    expect(onChange).toHaveBeenCalledTimes(1)
  })

  it('calls onBlur', () => {
    const onBlur = jest.fn()
    const component = shallow(<FilePicker onBlur={onBlur} />)
    const e = {
      target: {
        files: [{ name: 'data.json' }]
      }
    }

    component.find(`.${CLASS_PREFIX}-file-input`).simulate('change', e)
    component.find(`.${CLASS_PREFIX}-text-input`).simulate('blur')
    expect(onBlur).toHaveBeenCalledTimes(1)
  })

  it('sets placeholder', () => {
    const component = shallow(<FilePicker placeholder="placeholder here!" />)
    expect(component.find(`.${CLASS_PREFIX}-text-input`).prop('placeholder')).toEqual('placeholder here!')
  })
})
