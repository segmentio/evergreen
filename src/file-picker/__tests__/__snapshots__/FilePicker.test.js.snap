// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FilePicker /> snapshot 1`] = `
<div
  className="evergreen-file-picker-root ub-dspl_flex ub-box-szg_border-box"
>
  <input
    className="evergreen-file-picker-file-input ub-dspl_none ub-box-szg_border-box"
    onChange={[Function]}
    type="file"
  />
  <input
    aria-invalid={false}
    className="css-12rdf6g evergreen-file-picker-text-input ub-w_280px ub-fnt-fam_b77syt ub-b-btm_1px-solid-transparent ub-b-lft_1px-solid-transparent ub-b-rgt_1px-solid-transparent ub-b-top_1px-solid-transparent ub-otln_iu2jf4 ub-txt-deco_none ub-bblr_4px ub-bbrr_0-important ub-btlr_4px ub-btrr_0-important ub-ln-ht_16px ub-fnt-sze_12px ub-color_474d66 ub-pl_12px ub-pr_12px ub-tstn_n1akt6 ub-h_32px ub-bg-clr_white ub-b-btm-clr_d8dae5 ub-b-lft-clr_d8dae5 ub-b-rgt-clr_d8dae5 ub-b-top-clr_d8dae5 ub-flx_1 ub-txt-ovrf_ellipsis ub-f-wght_400 ub-ltr-spc_0 ub-box-szg_border-box"
    disabled={false}
    onBlur={[Function]}
    placeholder="Select a file to uploadâ€¦"
    readOnly={true}
    spellCheck={true}
    type="text"
    value=""
  />
  <button
    className="css-12b25td evergreen-file-picker-button ub-pst_relative ub-f-wght_400 ub-dspl_inline-flex ub-algn-itms_center ub-flx-wrap_nowrap ub-just-cnt_center ub-txt-deco_none ub-ver-algn_middle ub-b-btm_1px-solid-c1c4d6 ub-b-lft_1px-solid-c1c4d6 ub-b-rgt_1px-solid-c1c4d6 ub-b-top_1px-solid-c1c4d6 ub-otln_iu2jf4 ub-usr-slct_none ub-crsr_pointer ub-wht-spc_nowrap ub-fnt-fam_b77syt ub-bblr_0px ub-bbrr_4px ub-btlr_0px ub-btrr_4px ub-color_474d66 ub-tstn_n1akt6 ub-h_32px ub-min-w_32px ub-fnt-sze_12px ub-ln-ht_16px ub-pl_16px ub-pr_16px ub-bg-clr_white ub-flx-srnk_0 ub-ltr-spc_0 ub-box-szg_border-box"
    onBlur={[Function]}
    onClick={[Function]}
    type="button"
  >
    Select file
  </button>
</div>
`;
